Ghosts with hair.

Ricotez opened the pull request:
**This is NOT a direct port from Goonstation. Their ghosts inspired me, but I wrote the code from scratch. I didn't even look at Gooncode.**

**UPDATE:** You can now set two new preferences, Ghost Accessories and Ghosts of Others. The first one affects your world sprite lets you choose whether you want your ghost sprite to show hair and directional sprites, only directional sprites, or its original sprite. The second one only affects your client and lets you choose what the ghosts of others look like to you: their setting, their default sprites (without directions or hair, but still using the ghost form they've chosen) or as the simple white sprite.

Also, ghost hair is now a brighter version of its mortal color, to make ghosts look a bit less like aetherial lollypops. This has the added advantage of making the hair merge a bit better with the rest of the sprite.

![ghost6s](https://cloud.githubusercontent.com/assets/3974988/13998606/4440b0f0-f138-11e5-8785-e839a479340c.png)

**ORIGINAL POST**

If you have the default ghost form selected, the game will attempt to apply your head and facial hair to your ghost sprite. For this it reads the data of the body you occupied at the time of ghosting. If its species doesn't support head hair and/or facial hair, they won't show up, to prevent any sudden haircut disasters from preferences that you didn't know were set. So sorry lizards, you will get your turn another time! (By the way, if you observe, the game will take the data from the active character in your setup instead.)

**(These screenshots are outdated now, ghost hair is brighter.)**
![ghost1s](https://cloud.githubusercontent.com/assets/3974988/13901992/113c28c8-ee37-11e5-90df-eeb21989fbdb.png)
![ghost2s](https://cloud.githubusercontent.com/assets/3974988/13901993/128d5026-ee37-11e5-9a71-4c90e4096e54.png)
![ghost3s](https://cloud.githubusercontent.com/assets/3974988/13901994/1423ffac-ee37-11e5-9ca8-fcebd0a270f8.png)

This only works with the default ghost form (named "ghost" in the list, it's also the option the game forces you in if you don't have BYOND premium). I might extend it to the recoloured sprites eventually, but I don't want to force the hair on all ghost forms, because that would mean sprites like ianghost or ghostking become impossible (or have a haircut floating over them).

**(This screenshot is also outdated now.)**
![ghost4s](https://cloud.githubusercontent.com/assets/3974988/13901996/289f99d2-ee37-11e5-858c-25bb977408cb.png)

:cl:
imageadd: The default white ghost form now has directional sprites.
rscadd: The default white ghost form now also has (your) hair! Here's how it works.
experiment: If you ghost from a body, you'll get the hair from that body. If it has hair.
experiment: If you hit Observe, you'll get the hair from the character you had active in your setup window. If they had hair.
wip: This only works with the default white ghost form right now (because the other forms have the wrong shape or lack directional sprites), and with human bodies since they're the only ones that have hair. Sorry lizards and plasmamen! You'll get your turn next time.
rscadd: A new preference called Ghost Accessories. This lets you configure if you want your ghost sprite to show both hair and directional sprites, only directional sprites or just use its original default version without directional sprites.
rscadd: A new preference called Ghosts of Others. This preference is clientside and lets you configure how other ghosts appear to you: as their own setting, as the form they chose but without any hair or directional sprites, or as the simple white ghost sprite if you're sick of those premium users with their rainbow ghosts.
experiment: The layout of the ghost preference buttons has slightly changed. The old Choose Ghost Form and Choose Ghost Orbit are now a part of Ghost Customization, which also contains the new Ghost Accessories button. Clicking Ghost Customization will send you directly to Ghost Accessories if you don't have premium. The Preferences tab now also contains the Ghost Display button, which lets you configure how other ghosts appear to you.
/:cl:

Ricotez committed:
+Added directional sprites to iconstate ghost in mob.dmi
+Ghosts now update their dir variable when moving around

Ricotez committed:
+When the default ghost sprite is selected, and your species has head and/or facial hair, your ghost sprite will have the same hair.
+If you ghost in the round, this will base itself on the body you occupied.
+If you observe, this will base itself on the character you had active when you hit the button.

Lobachevskiy commented:
Blackwell when

PKPenguin321 commented:
:+1: super awesome

RemieRichards commented:
Knew it was this that you needed the ghost help with for :P 
:+1: 

As for hair, why not make it a pref? that way people can enable/disable it for silly things like Ian with an afro.

Zilenan91 commented:
it's cool

Incoming5643 commented:
I don't like giving attention to ghosts, it distracts from the things that actually matter

RemieRichards commented on code/modules/mob/dead/observer/observer.dm:
```diff
@@ -112,6 +175,7 @@ This is the proc mobs get to turn into a ghost. Forked from ghostize due to comp
 
 
 /mob/dead/observer/Move(NewLoc, direct)
+	dir = direct //For the new fancy directional sprites. |- Ricotez
```
this comment isn't really necessary

duncathan replied:
Also signed comments are a no-no

RemieRichards replied:
yeah @duncathan I was going to make that point (I actually did further above) but realised the whole comment is useless so it can all be dropped. but ye.

RemieRichards commented on code/modules/mob/dead/observer/observer.dm:
```diff
@@ -77,6 +96,50 @@ var/list/image/ghost_darkness_images = list() //this is a list of images for thi
+		icon_state = new_form
+		ghostimage.icon_state = new_form
+
+	if(icon_state == "ghost") //For now we just check for the default ghost sprite. If other sprites get updated in the future to support hair, change this to a list. |- Ricotez
```
remove the |- Ricotez from this one, the one on the big banner comment to stay.
generally 1 "I did this" comment is allowed on bigger things like procs or frameworks.

duncathan replied:
Oh I guess remie already said it

RemieRichards replied:
:P

RemieRichards commented:
@Incoming5643 good job you're not the one who decided to do this then eh? doesn't hurt you.

PKPenguin321 commented:
>I don't like giving attention to ghosts, it distracts from the things that actually matter

then turn off ghost vision

Incoming5643 commented:
I do like watching ghosts, I just don't like them being distracting. It's just like, an opinion man

Ricotez committed:
-Small modifications to the comments.

Shapsy commented:
Finally I have some way of identifying the shitter ghost who flies away from my body just as I die :+1:

Cheridan commented:
it looks funny, you can see the outline of the ghost sprite on top of the hair

Ricotez commented:
Goonstation has the exact same problem, and there's not really an easy fix for it, at least not one I can think of. Just a downside of making everything semitransparent.

Personally I think it makes the hair look more spooky, like some kind of aethereal halo.

YotaXP commented:
This is actually a feature that I've always hated on other codebases.  They always look too creepy IMO.  Also I kinda like floating around without thinking about people going "Hey, look!  It's the ghost of THAT guy!  What a dick."  I'd rather blend in with all the other ghosts.

Niknakflak commented:
I personally do not like this.

optimumtact commented:
Make this a byond donator feature please we need to encourage people to donate

Steelpoint commented:
Niknak for disliking most cool things 2k16

Otherwise this is a nice feature to get that many other servers use in some capacity. From Goon's ghosts to others just taking a snapshot of your body on death and making that the ghost form.

Steelpoint commented:
Also if you think this makes Ghosts more distracting, there's always the option of hiding ghosts altogether.

Iamgoofball commented:
I might swap our ghost for the goon ghost sprite, it's better for this

ChangelingRain commented:
I don't like this, I prefer ghost anonymity and would prefer there be an option to turn this off.

Wyzack commented:
>Complaining that ghosts are too scary

Is this a real complaint? Seems more like testimony that they should be added

Iamgoofball commented:
@ChangelingRain but it already shows names if you hover over it

ChangelingRain commented:
Goof I don't give a shit I prefer it when all the ghosts look the same

MMMiracles commented:
i don't understand why you'd want to be 'anonymous' as a ghost if you're literally that scared of metagrudge or some shit just use a random character.

ChangelingRain commented:
At minimum I want to be able to turn off the directional sprites and hair so I don't have to look at them but can still see ghosts.

Supermichael777 commented:
such spookiness much scared

Shadowlight213 commented:
Yeah, this probably needs toggles to hide viewing others' ghost hair and maybe toggles for your own ghost hair too.

Ricotez commented:
@Shadowlight213 I've been trying to figure out a way to achieve the toggle on other people's ghost sprites, but it's not going to be easy. There's probably some simple way to achieve it that I haven't thought of yet though.

Shadowlight213 commented:
@ricotez try using drone vision filter code. Except instead of having it be an outline, have it override the sprite to a ghost sprite without the hair. If possible, make it respect the ghost alternate sprites for byond members still.

Bawhoppen commented:
:-1: it looks different and weird

duncathan commented:
:-1:

MMMiracles commented:
ITT: ghost hair too spoopy for some people

Ricotez commented:
I'm nearly done with adding preferences for this stuff, just have to add in the client-side sprite filter. There are going to be separate preferences for your own sprite (full accessories, only new directional sprites, classic mode) and for what other ghosts look like to you (their setting, classic mode, all simple white ghosts).

Steelpoint commented:
I like this because it seem like a unique way of showing off ghosts.

Other servers either go for the Goon method or a straight copy of the dead person as a ghost. This one is a unique interpretation

Also its pretty cool.

KorPhaeron commented:
I hate change

duncathan commented:
if I can make it so that I don't see other ghost's hair then :+1:

Ricotez commented:
@duncathan The display preferences go one step further. You can even completely disable custom ghost forms, which will make everyone else show up as the default white ghost sprite without accessories or directional sprites regardless of what their settings are.

Ricotez committed:
Brighter ghost hair and new preferences.
+Ghost hair is now brighter than before.
+Added a new preference ghost_accs that stores what you want your own ghost to display.
+Added a new preference ghost_others that stores what you want other people's ghosts to display.

Ricotez commented:
I've finally finished debugging all the new stuff. Newest commit contains preference options for your own and for other people's ghosts, and ghost hair is now a brighter version of its original. For full details, refer to the (updated) OP.

I think I've covered the majority of the complaints about the PR with this update, so I hope everyone's okay with the merge now.

optimumtact commented:
Talk about going the extra mile :heart:

YotaXP commented:
You've turned my :-1: into a :+1:

:clap: :clap: :clap: :tada:

Ricotez commented:
@RemieRichards @Cheridan @ChangelingRain @duncathan @KorPhaeron It's been 24 hours since the last commit, any news on whether this'll get merged? Or do you want me to change anything else?

KorPhaeron commented:
@Ricotez 

I'd merge it if I felt comfortable reviewing the code, it seems fine

ChangelingRain commented:
What's the default for those ghost options?

Ricotez commented:
@ChangelingRain Defaults for Ghost Accessories is Full Accessories, default for Ghosts of Others is Their Settings. If you change the settings in the setup window, you can save your preferences so it remembers them next time.
